### PR TITLE
Type erasing ResponseBodyWriter

### PR DESCRIPTION
**Pull request by adam-fowler**
- Rename ResponseBodyWriter to ResponseBodyWriterProtocol
- Add type ResponseBodyWriter that holds an existential ResponseBodyWriterProtocol
- Broke code into three files to make it easier to read

### CODE REVIEW

**Review by sliemeobn — COMMENTED** [review]
nice!

I think this provides a nicer API for "users" of `ReponseBody` and more ergonomic extensions of it (ie: `ResponseBodyWriter` can now conform to protocols)

not 100% sure I like the naming of `ResponseBodyWriterProtocol`. this will only affect implementers of a ResponseBodyWriter, but it might be potentially confusing.

in the olden days words like `ResponseBodyWriting` were used, but it seems this is falling out of favor 🤔

**Review by Joannis — CHANGES_REQUESTED** [review]
I understand the desire for this change, and I like it a lot. But I don't think that the value provided by this change outweighs a breaking change so late in the RC phase.

I'm happy to be convinced otherwise, maybe if more people weigh in

**Comment by adam-fowler** [comment]
Another option is I ditch all the API breaking changes and simply add a `AnyResponseBodyWriter` type which the user can use in their own code.

File: Sources/HummingbirdCore/Response/ResponseBodyWriter.swift [hunk 1]
```diff
@@ -15,67 +15,24 @@
...
-    mutating func write(contentsOf buffers: some Sequence<ByteBuffer>) async throws
-    /// Finish writing body
-    /// - Parameter trailingHeaders: Any trailing headers you want to include at end
-    consuming func finish(_ trailingHeaders: HTTPFields?) async throws
-}
+/// Type erasing ``ResponseBodyWriterProtocol``.
+///
+/// Holds a pointer to any ResponseBodyWriterProtocol.
+public struct ResponseBodyWriter: ResponseBodyWriterProtocol {
+    public var wrapped: any ResponseBodyWriterProtocol
```
Comment (MahdiBM) [inline Sources/HummingbirdCore/Response/ResponseBodyWriter.swift:22]: I wonder if sometime in the future you'd want to make an special case for hummingbird's own response-body-writer with an enum, for performance optimizations.

Just something to keep in mind. This public var shouldn't stop use from doing that.

Comment (MahdiBM) [reply Sources/HummingbirdCore/Response/ResponseBodyWriter.swift:22]: (Like `NIOAny`, or even String's own "guts")

**Comment by adam-fowler** [comment]
Both protocol and type erased type need to be available. Users need to be able to write their own Response body writers. This new type fulfills the need for a concrete type that can be extended.

**Comment by adam-fowler** [comment]
I've decided to not go ahead with this PR.

**Pull request closed without merging**